File: android/src/main/kotlin/com/example/xplayer/PlayerViewController.kt
package com.example.xplayer

import android.content.Context
import android.util.Log
import androidx.media3.ui.PlayerView
import io.flutter.plugin.common.MethodCall

class PlayerViewController(private var context: Context) {
    private var playerViews = mutableMapOf<String, PlayerView>()
    private var currentPlayerViewId: String = "";
    fun getCurrentView(): PlayerView? {
        try {
            if (playerViews.isEmpty()) return null

            return playerViews[currentPlayerViewId]

        } catch (e: Exception) {
            Log.d("XPlayer", "Failed to get current PlayerView $e")
            return null;
        }
    }

    fun registerView(id: String, view: PlayerView){
        playerViews[id] = view;
    }

    fun removeView(call: MethodCall){
        val viewId = call.arguments as String? ?: return
        playerViews.remove(viewId)
    }

    fun getCurrentPlayerViewId(): String {
        return currentPlayerViewId
    }

    fun setCurrentPlayerViewId(id: String) {
         currentPlayerViewId = id
    }

    fun getPlayerViewById(id: String): PlayerView?{
        return playerViews[id];
    }
}

File: ios/Classes/PlayerViewController.swift
//
//  PlayerViewController.swift
//  xplayer
//
//  Created by Sovanreach on 13/1/24.
//

import Flutter
import Foundation
import AVFoundation

class PlayerViewController: NSObject {
    private var playerViews: [String : PlayerView] = [:]
    private var currentPlayerViewId: String = ""
    
    func getCurrenView() -> UIView? {
        
        if(playerViews.isEmpty) { return nil }
        
        guard let playerView = playerViews[currentPlayerViewId] else{
            return nil
        }
        
        return playerView
    }
    
    func registerView(id: String, view: PlayerView){
        playerViews[id] = view
    }
    
    func remove(call: FlutterMethodCall, result: FlutterResult){
        guard let viewId = call.arguments as? String else{
            result("Arguments sent from Flutter is not String")
            return
        }
        playerViews.removeValue(forKey: viewId)
    }
    
    func getCurrentPlayerViewId() -> String {
        return currentPlayerViewId
    }
    
    func setCurrentPlayerViewId(id: String) {
        currentPlayerViewId = id
    }
    
    func getPlayerViewById(id: String) -> PlayerView? {
        return playerViews[id]
    }
    
    /// Special case for iOS,
    /// If you check Android implementation, you will see they have built-in switching player ```PlayerView.switchTargetView(player, oldPlayerView, newPlayerView)```
    /// This is manual switching player between oldPlayerView and newPlayerView
//    func switchTargetView(player: AVPlayer?, oldPlayerView: UIView?, newPlayerView: UIView?){
//        if(player == nil) {return}
//        
//
//        
//        if(newPlayerView != nil){
//            (oldPlayerView?.layer as! AVPlayerLayer).player = player
//        }
//        
//        if(oldPlayerView != nil) {
//            (oldPlayerView?.layer as! AVPlayerLayer).player = nil
//        }
//    }
}

